bundles/org.openhab.binding.wundergroundupdatereceiver/src/main/java/org/openhab/binding/wundergroundupdatereceiver/internal/WundergroundUpdateReceiverHandler.java

```diff
@@ -0,0 +1,176 @@
+import org.openhab.core.types.Command;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link WundergroundUpdateReceiverHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Daniel Demus - Initial contribution
+ */
+@NonNullByDefault
+public class WundergroundUpdateReceiverHandler extends BaseThingHandler {
+
+    public String getStationId() {
+        return config.stationId;
+    }
+
+    private final Logger logger = LoggerFactory.getLogger(WundergroundUpdateReceiverHandler.class);
+    private final WundergroundUpdateReceiverServlet wundergroundUpdateReceiverServlet;
+
+    private final ChannelUID lastReceivedChannel;
+    private final ChannelUID queryStateChannel;
+    private final ChannelUID queryTriggerChannel;
+
+    private WundergroundUpdateReceiverConfiguration config = new WundergroundUpdateReceiverConfiguration();
+
+    public WundergroundUpdateReceiverHandler(Thing thing,
+            WundergroundUpdateReceiverServlet wunderGroundUpdateReceiverServlet,
+            WundergroundUpdateReceiverDiscoveryService discoveryService,
+            WundergroundUpdateReceiverUnknownChannelTypeProvider channelTypeProvider,
+            ChannelTypeRegistry channelTypeRegistry) {
+        super(thing);
+        this.lastReceivedChannel = new ChannelUID(new ChannelGroupUID(getThing().getUID(), METADATA_GROUP),
+                LAST_RECEIVED_DATETIME);
+        this.queryTriggerChannel = new ChannelUID(new ChannelGroupUID(getThing().getUID(), METADATA_GROUP),
+                LAST_QUERY_TRIGGER);
+        this.queryStateChannel = new ChannelUID(new ChannelGroupUID(getThing().getUID(), METADATA_GROUP),
+                LAST_QUERY_STATE);
+        String stationId = thing.getProperties()
+                .get(WundergroundUpdateReceiverBindingConstants.REPRESENTATION_PROPERTY);
+        if (thing.getChannels().isEmpty()) {
+            ThingBuilder thingBuilder = editThing();
+            @Nullable
+            Map<String, String[]> requestParameters = discoveryService.getUnhandledStationRequest(stationId);
+            if (requestParameters != null) {
+                requestParameters.forEach((String parameter, String[] query) -> {
+                    WundergroundUpdateReceiverParameterMapping channelTypeMapping = WundergroundUpdateReceiverParameterMapping
+                            .getOrCreateMapping(parameter, String.join("", query), channelTypeProvider);
+                    ChannelBuilder channelBuilder = ChannelBuilder.create(
+                            new ChannelUID(thing.getUID(), channelTypeMapping.channelGroup, parameter),
+                            channelTypeRegistry.getChannelType(channelTypeMapping.channelTypeId).getItemType());
+                    thingBuilder.withChannel(channelBuilder.build());
+                });
+            }
+            updateThing(thingBuilder.build());
+        }
```

Comment:
If I understand correctly this is done exactly once, after the Thing has been discovered. It seems to make sense to do this during the discovery process.

Reply:
As far as I can tell you only have ThingUID in the discovery service. The thing is created right before initialization, but of course I could be mistaken.
Especially now that ad-hoc addition of extra channels received in requests has been added to support multiple devices submitting to the same WU account, it makes sense to have it here.